File: lib/python/opentoken/src/main/opentoken/io/parquet/person_attributes_parquet_writer.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import logging
import os
from typing import Dict, Optional
try:
    import pyarrow as pa
    import pyarrow.parquet as pq
except ImportError:
    raise ImportError("pyarrow is required for Parquet support. Install with: pip install pyarrow")

from opentoken.io.person_attributes_writer import PersonAttributesWriter


logger = logging.getLogger(__name__)


class PersonAttributesParquetWriter(PersonAttributesWriter):
    """
    The PersonAttributeParquetWriter class is responsible for writing person
    attributes to a Parquet file.
    It implements the PersonAttributesWriter interface.
    """

    def __init__(self, file_path: str):
        """
        Initialize the class with the output file in Parquet format.

        Args:
            file_path: The output file path.

        Raises:
            IOError: If an I/O error occurs.
        """
        self.file_path = file_path

        # Create directory if it doesn't exist
        os.makedirs(os.path.dirname(file_path), exist_ok=True)

        self.schema: Optional[pa.Schema] = None
        self.writer: Optional[pq.ParquetWriter] = None
        self.initialized = False

    def write_attributes(self, attributes: Dict[str, str]) -> None:
        """
        Write attributes to the Parquet file.

        Args:
            attributes: A map of person attributes.

        Raises:
            IOError: If an I/O error occurs.
        """
        if not self.initialized:
            self._initialize_writer(attributes)

        # Create a record batch from the attributes
        arrays = []
        field_names = []

        for field in self.schema:
            field_name = field.name
            field_names.append(field_name)
            value = attributes.get(field_name)
            if value is not None:
                arrays.append(pa.array([value]))
            else:
                arrays.append(pa.array([None], type=pa.string()))

        record_batch = pa.record_batch(arrays, names=field_names)
        self.writer.write_batch(record_batch)

    def close(self) -> None:
        """Close the Parquet writer."""
        if self.writer is not None:
            self.writer.close()

    def _initialize_writer(self, first_record: Dict[str, str]) -> None:
        """
        Initialize the Parquet writer with schema based on the first record.

        Args:
            first_record: The first record to determine schema.

        Raises:
            IOError: If an I/O error occurs.
        """
        # Create schema based on the first record
        fields = []
        for key, value in first_record.items():
            if value is not None:
                fields.append(pa.field(key, pa.string()))

        self.schema = pa.schema(fields)

        # Initialize the Parquet writer
        self.writer = pq.ParquetWriter(self.file_path, self.schema)
        self.initialized = True

File: lib/java/opentoken/src/main/java/com/truveta/opentoken/io/parquet/PersonAttributesParquetWriter.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io.parquet;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.parquet.hadoop.ParquetWriter;
import org.apache.parquet.hadoop.ParquetFileWriter;
import org.apache.parquet.hadoop.example.ExampleParquetWriter;
import org.apache.parquet.hadoop.example.GroupWriteSupport;
import org.apache.parquet.column.ParquetProperties.WriterVersion;
import org.apache.parquet.example.data.Group;
import org.apache.parquet.example.data.simple.SimpleGroup;
import org.apache.parquet.schema.MessageType;
import org.apache.parquet.schema.MessageTypeParser;

import com.truveta.opentoken.io.PersonAttributesWriter;
import java.io.IOException;
import java.util.Map;

/**
 * The PersonAttributeParquetWriter class is responsible for writing person
 * attributes to a Parquet file.
 * It implements the {@link PersonAttributeWriter} interface.
 */
public class PersonAttributesParquetWriter implements PersonAttributesWriter {
    private ParquetWriter<Group> writer;
    private MessageType schema;
    private final String filePath;
    private final Configuration conf;
    private boolean initialized = false;

    /**
     * Initialize the class with the output file in Parquet format.
     * 
     * @param filePath the output file path
     * @throws IOException if an I/O error occurs
     */
    public PersonAttributesParquetWriter(String filepath) throws IOException {
        this.filePath = filepath;
        this.conf = new Configuration();
    }

    @Override
    public void writeAttributes(Map<String, String> attributes) throws IOException {
        if (!initialized) {
            initializeWriter(attributes);
        }

        SimpleGroup group = new SimpleGroup(schema);
        attributes.forEach((key, value) -> {
            if (schema.containsField(key) && value != null) {
                group.add(key, value);
            }
        });

        writer.write(group);
    }

    @Override
    public void close() throws IOException {
        if (writer != null) {
            writer.close();
        }
    }

    private void initializeWriter(Map<String, String> firstRecord) throws IOException {
        StringBuilder schemaBuilder = new StringBuilder();
        schemaBuilder.append("message Person {\n");

        firstRecord.forEach((key, value) -> {
            if (value != null) {
                schemaBuilder.append("  required binary ").append(key).append(" (UTF8);\n");
            }
        });

        schemaBuilder.append("}");

        this.schema = MessageTypeParser.parseMessageType(schemaBuilder.toString());
        GroupWriteSupport.setSchema(schema, conf);
        Path path = new Path(filePath);

        writer = ExampleParquetWriter.builder(path)
                .withWriteMode(ParquetFileWriter.Mode.OVERWRITE)
                .withWriterVersion(WriterVersion.PARQUET_2_0)
                .withConf(conf)
                .withType(schema)
                .build();

        initialized = true;
    }
}
